# Exercise 4.9 Task Part 1

## 1. Import libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## 2. Set path

In [3]:
# Set system path
path = r'C:\Users\Stony\OneDrive\CareerFoundry\Data Immersion Course\Instacart Basket Analysis'

## 3. Import customer data into a dataframe

In [5]:
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

## 4. Validate new customer dataframe parameters

In [6]:
# Check dataframe headers
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [7]:
# Check shape of customers dataframe
customers.shape

(206209, 10)

In [8]:
# Check description of customers dataframe
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [9]:
# Check info of customers dataframe
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


## 5. Wrangle data to follow consistent logic

Currently, all other data is lower case letters only, with underscores instead of spaces

In [11]:
# Rename First Name to first_name
customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [12]:
# Rename Surnam to last_name
customers.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [13]:
# Rename Gender to gender
customers.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [14]:
# Rename STATE to state
customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [15]:
# Rename Age to age
customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [18]:
# Rename fam_status to marital_status
customers.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [19]:
# Validate the name changes were successful
customers.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'n_dependants', 'marital_status', 'income'],
      dtype='object')

6 columns were renamed to be more clear to their contents and to follow the logic used in prior dataframes. I do not see any columns that are not needed currently, so no columns will be dropped at this time.

## 6. Data consistency checks on customers dataframe

In [20]:
# Check for missing values
customers.isnull().sum()

user_id               0
first_name        11259
last_name             0
gender                0
state                 0
age                   0
date_joined           0
n_dependants          0
marital_status        0
income                0
dtype: int64

There appear to be 11,259 rows that are missing the first name.

In [22]:
# Create a new dataframe called df_dups to check for duplicate values
df_dups = customers[customers.duplicated()]

In [23]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, marital_status, income]
Index: []

There are no duplicates in the customers dataframe

In [24]:
# Check for mixed-type columns
customers.dtypes

user_id            int64
first_name        object
last_name         object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [30]:
for col in customers.columns.tolist():
    weird = (customers[[col]].map(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers[weird]) > 0:
        print (col)

first_name


Our first_name column contains mixed data in it.

In [32]:
# Convert first_name column to string data type
customers['first_name'] = customers['first_name'].astype('str')

In [33]:
# Validate data types 
customers.dtypes

user_id            int64
first_name        object
last_name         object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [37]:
# Validate there are no longer any mixed data-type columns in the customers dataframe
for col in customers.columns.tolist():
    weird = (customers[[col]].map(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers[weird]) > 0:
        print (col)

There are no longer any columns that contain mixed data-types in the customer dataframe

## 7. Combine customer dataframe with the rest of our prepared Instacart data

In [40]:
# Import the ords_prods_derived dataframe
ords_prods_derived = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_derived.pkl'))

In [41]:
# Check data types in ords_prods_derived dataframe
ords_prods_derived.dtypes

product_id                       int64
product_name                    object
aisle_id                         int64
department_id                    int64
prices                         float64
order_id                         int64
user_id                          int64
order_number                     int64
order_day_of_week                int64
order_hour_of_day                int64
days_since_last_order          float64
first_order                       bool
add_to_cart_order                int64
reordered                        int64
_merge                        category
price_range_loc                 object
busiest_day                     object
busiest_days                    object
busiest_period_of_day           object
max_order                        int64
loyalty_flag                    object
mean_product_price             float64
spending_flag                   object
median_days_between_orders     float64
order_frequency                 object
dtype: object

In [42]:
# Drop the _merge column from the dataframe
ords_prods_derived = ords_prods_derived.drop(columns=['_merge'])

In [43]:
# Merge customers and ords_prods_derived dataframes together on user_id column
ords_prods_customers_combined = ords_prods_derived.merge(customers, on = 'user_id', indicator = True)

In [44]:
# Validate merge was successful
ords_prods_customers_combined.dtypes

product_id                       int64
product_name                    object
aisle_id                         int64
department_id                    int64
prices                         float64
order_id                         int64
user_id                          int64
order_number                     int64
order_day_of_week                int64
order_hour_of_day                int64
days_since_last_order          float64
first_order                       bool
add_to_cart_order                int64
reordered                        int64
price_range_loc                 object
busiest_day                     object
busiest_days                    object
busiest_period_of_day           object
max_order                        int64
loyalty_flag                    object
mean_product_price             float64
spending_flag                   object
median_days_between_orders     float64
order_frequency                 object
first_name                      object
last_name                

In [45]:
## 8. Export combined dataframe as a pickle file
ords_prods_customers_combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers_combined.pkl'))